File: RotateImage.java
/*
Runtime: 0 ms, faster than 100.00% of Java online submissions for Rotate Image.
Memory Usage: 43.1 MB, less than 30.35% of Java online submissions for Rotate Image.
*/
/*
    | TC: O(n) (n = Number of cells in the grid) | SC: O(1) |
    
    Explanation of Time Complexity:
    1. Transposing the matrix has a cost of O(n)(we're moving the value of each cell once).
    2. Reversing each row also has a cost of O(n) (we're moving the value of each cell once).


*/
class Solution {
    public void rotate(int[][] matrix) {
        
        int numRows = matrix.length;
        
        // Transpose The Matrix
        for(int i=0; i<numRows; i++){
            for(int j=i; j<numRows; j++){
                
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
                
            }
        }
        
        // Swap The Elements From Both Side Moving Towards The Centre
        for(int i=0; i<numRows; i++){
            for(int j = 0; j<(numRows/2); j++){
                
                int temp = matrix[i][j];
                matrix[i][j] = matrix[i][numRows - 1 - j];
                matrix[i][numRows - 1 - j] = temp;
                
            }
        }
    }
}

File: RotateImage.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
// TC: O(n^2) SC: O(1)
var rotate = function(matrix) {
    const reverse = (nums) => {
        let start = 0, end = nums.length - 1;
        while(start < end) {
            const temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
    for(let i=0; i<matrix.length; i++) {
        for(let j=i; j<matrix[0].length; j++) {
            const temp = matrix[i][j];
            matrix[i][j] = matrix[j][i];
            matrix[j][i] = temp;
        }
        
    }
    for(let i=0; i<matrix.length; i++) {
        reverse(matrix[i]);
    }
};
